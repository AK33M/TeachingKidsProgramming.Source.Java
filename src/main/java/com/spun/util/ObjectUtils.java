package com.spun.util;

import org.teachingkidsprogramming.util.ArrayUtils;
import org.teachingkidsprogramming.util.ClassUtils;

import java.lang.reflect.Array;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * A static class of convenience functions for Manipulating objects
 **/
public class ObjectUtils {
	public static Method[] getMethodsForObject(Object o2, String[] passedMethods)
			throws SecurityException, NoSuchMethodException {
		Method methods[] = new Method[passedMethods.length];
		Class clazz = o2.getClass();
		for (int i = 0; i < passedMethods.length; i++) {
			methods[i] = clazz.getMethod(passedMethods[i], (Class[]) null);
		}
		return methods;
	}

	/**
	 * A convenience function to check if two objects are equal.
	 * 
	 * @param s1
	 *            The first object
	 * @param s2
	 *            The second object
	 * @return true if Equal.
	 **/
	public static boolean isEqual(Object s1, Object s2) {
		if (s1 == s2) {
			return true;
		} else if ((s1 != null) && s1.equals(s2)) {
			return true;
		} else {
			return false;
		}
	}

	public static boolean isIn(Object target, Object[] objects) {
		for (int i = 0; i < objects.length; i++) {
			if (ObjectUtils.isEqual(objects[i], target)) {
				return true;
			}
		}
		return false;
	}

	public static boolean isThisInstanceOfThat(Class<?> thiz, Class<?> that) {
		return that.isAssignableFrom(thiz);
	}

	public static Error throwAsError(Throwable t) throws Error {
		if (t instanceof RuntimeException) {
			throw (RuntimeException) t;
		} else if (t instanceof Error) {
			throw (Error) t;
		} else {
			throw new Error(t);
		}
	}

	/**
	 * @deprecated use Query.select()
	 */
	public static Object[] extractArray(Object[] from, String methodName) {
		try {
			if (from == null || from.length == 0) {
				return new Object[0];
			}
			Method method = getGreatestCommonDenominator(from, methodName);
			Object[] array = null;
			if (Object.class.isAssignableFrom(method.getReturnType())) {
				array = (Object[]) Array.newInstance(method.getReturnType(),
						from.length);
			} else {
				array = (Object[]) Array.newInstance(
						ClassUtils.getWrapperClass(method.getReturnType()),
						from.length);
			}
			for (int i = 0; i < from.length; i++) {
				array[i] = method.invoke(from[i], (Object[]) null);
			}
			return array;
		} catch (Exception e) {
			MySystem.warning(e);
			throw ObjectUtils.throwAsError(e);
		}
	}

	public static Method getGreatestCommonDenominator(Object[] from,
			String methodName) throws SecurityException, NoSuchMethodException {
		List<Class> classes = new ArrayList<Class>();
		ArrayUtils.addArray(classes, getAllCastableClasses(from[0]));
		for (Object o : from) {
			for (int i = classes.size() - 1; i >= 0; i--) {
				Class clazz = classes.get(i);
				if (!isThisInstanceOfThat(o.getClass(), clazz)
						|| !ClassUtils.hasMethod(clazz, methodName)) {
					classes.remove(i);
				}
			}
		}
		return classes.size() == 0 ? null : ArrayUtils.getLast(classes)
				.getMethod(methodName, (Class[]) null);
	}

	private static Class[] getAllCastableClasses(Object object) {
		Class<? extends Object> clazz = object.getClass();
		ArrayList<Object> list = new ArrayList<Object>();
		while (clazz != null) {
			list.add(clazz);
			ArrayUtils.addArray(list, clazz.getInterfaces());
			clazz = clazz.getSuperclass();
		}
		Class[] found = (Class[]) list.toArray(new Class[list.size()]);
		ArrayUtils.toReverseArray(found);
		return found;
	}

	public static void assertInstance(Class clazz, Object object) {
		assertInstance(new Class[] { clazz }, object);
	}

	public static void assertInstance(Class classes[], Object object) {
		if (object == null) {
			throw new NullPointerException("Expected Object of Type "
					+ Arrays.asList(extractArray(classes, "getName"))
					+ " but was null");
		}
		for (int i = 0; i < classes.length; i++) {
			if (ClassUtils.getWrapperClass(classes[i]).isInstance(object)) {
				return;
			}
		}
		throw new IllegalArgumentException("Expected Object of Type "
				+ Arrays.asList(extractArray(classes, "getName")) + " but got "
				+ object.getClass().getName());
	}

	public static String getClassName(Object o) {
		return o == null ? "null" : o.getClass().getName();
	}
}