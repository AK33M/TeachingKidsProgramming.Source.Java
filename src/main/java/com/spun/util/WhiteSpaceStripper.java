package com.spun.util;

import org.teachingkidsprogramming.util.MySystem;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

public class WhiteSpaceStripper 
{
	/***********************************************************************/
	public static void stripFolder(File dir)
  {
    stripFolder(dir, true);
  }
	/***********************************************************************/
	public static void stripFolder(File dir, boolean recursive)
  {
  	if (!dir.isDirectory())
    {
    	MySystem.warning("File is not a Directory - " + dir.toString());
    	return;
    }
  	
  	File[] files = dir.listFiles(new WhiteSpaceFileFilter());
  	for (int i = 0; i < files.length; i++)
    {
    	if (files[i].isDirectory())
      {
      	MySystem.event("Scaning Directory -" + files[i].getName());
      	if (recursive) stripFolder(files[i], recursive);
      }
    	else
      {
        stripFile(files[i]);
      }
    }
  }

    /***********************************************************************/
	public static void stripFile(File file)
  {
    
  	if (!file.isFile())
    {
    	MySystem.warning("File is not a File - " + file.toString());
    	return;
    }
		if(!file.canWrite())
		{
			MySystem.event("File '" + file.toString() + "' is readonly");
        }
		else
		{
	    try
	    {
	      String contents = readFile(file);
	      String stripped = stripWhiteSpace(contents);
	      writeFile(file, stripped);
	    }
	    catch (IOException e)
	    {
	      MySystem.warning(e);
	    }
		}
  }
  /***********************************************************************/
  public static String stripWhiteSpace(String text)
  {
    StringBuilder newText = new StringBuilder();
    boolean whitespace = false;
    int num = text.length();
		char whiteSpaceChar = ' ';
    for (int i = 0; i < num; i++)
    {
      char c = text.charAt(i);
			switch (c)
			{
				case '\n' : whiteSpaceChar = '\n';whitespace = true;break;
				case '\t' :
				case ' '  : whitespace = true;break;
				default	  : 
					if (whitespace)
					{
						whitespace = false;
						newText.append(whiteSpaceChar);
						whiteSpaceChar = ' '; 
					}						
	        newText.append(c);break;
      }
    }
    return newText.toString();
  }

    /***********************************************************************/
  private static String readFile(File file)
  throws IOException
  {
    BufferedReader reader = new BufferedReader(new FileReader(file));
    StringBuilder output = new StringBuilder();
    
    while (reader.ready())
    {
      output.append(reader.readLine());
	    output.append("\n");
    }
    
    reader.close();
    return output.toString();
  }
	/***********************************************************************/
	private static void writeFile(File file, String text)
	throws IOException
	{
    FileWriter writer = new FileWriter(file);
    
    writer.write(text);
    
    writer.close();
	}
	/***********************************************************************/
  public static void main (String [] args)
  {
//    stripFolder(new File("C:\\temp\\stockgazing"));
  }
	/***********************************************************************/
  /***********************************************************************/
}


class WhiteSpaceFileFilter
  implements java.io.FileFilter
{

  /***********************************************************************/

  public boolean accept(File pathname)
  {
  	
    if (pathname.getName().equals(".") || pathname.getName().equals("."))
    {
      return false;
    }
    else if (pathname.getName().equalsIgnoreCase("email"))
    {
      return false;
    }
    else
        return pathname.isDirectory() || (pathname.getName().indexOf(".htm") != -1) || (pathname.getName().indexOf(".txt") != -1);
  }
  
  /***********************************************************************/
  /***********************************************************************/
}