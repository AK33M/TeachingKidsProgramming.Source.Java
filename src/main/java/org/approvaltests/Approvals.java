package org.approvaltests;

import java.awt.Component;
import java.awt.Image;
import java.awt.image.BufferedImage;
import java.io.File;
import java.util.Arrays;
import java.util.Map;

import org.approvaltests.ReporterFactory.FileTypes;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.ApprovalFailureOverrider;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.JUnitStackTraceNamer;
import org.approvaltests.writers.ApprovalTextWriter;
import org.approvaltests.writers.ApprovalXmlWriter;
import org.approvaltests.writers.ComponentApprovalWriter;
import org.approvaltests.writers.FileApprovalWriter;
import org.approvaltests.writers.ImageApprovalWriter;
import org.lambda.functions.Function1;

import org.teachingkidsprogramming.util.ArrayUtils;
import org.teachingkidsprogramming.util.ObjectUtils;
import com.spun.util.StringUtils;

public class Approvals {
	public static void verify(String response) throws Exception {
		verify(new ApprovalTextWriter(response, "txt"), FileTypes.Text);
	}

	public static void verify(Object o) throws Exception {
		verify("" + o);
	}

	public static <T> void verifyAll(String label, T[] array) throws Exception {
		verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"),
				FileTypes.Text);
	}

	public static <T> void verifyAll(String header, String label, T[] array)
			throws Exception {
		verify(new ApprovalTextWriter(formatHeader(header)
				+ StringUtils.toString(label, array), "txt"), FileTypes.Text);
	}

	public static <T> void verifyAll(T[] values, Function1<T, String> f1) {
		String text = ArrayUtils.toString(values, f1);
		verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
	}

	public static <T> void verifyAll(String header, T[] values,
			Function1<T, String> f1) {
		verifyAll(header, Arrays.asList(values), f1);
	}

	public static <T> void verifyAll(String header, Iterable<T> array,
			Function1<T, String> f1) {
		String text = formatHeader(header) + ArrayUtils.toString(array, f1);
		verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
	}

	private static String formatHeader(String header) {
		return StringUtils.isEmpty(header) ? "" : header + "\r\n\r\n\r\n";
	}

	public static <T> void verifyAll(String label, Iterable<T> array)
			throws Exception {
		verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"),
				FileTypes.Text);
	}

	public static <T> void verifyAll(String header, String label,
			Iterable<T> array) throws Exception {
		verify(new ApprovalTextWriter(formatHeader(header)
				+ StringUtils.toString(label, array), "txt"), FileTypes.Text);
	}

	public static void verify(Component c) throws Exception {
		verify(new ComponentApprovalWriter(c), FileTypes.Image);
	}

	public static void verifyHtml(String response) throws Exception {
		verify(new ApprovalTextWriter(response, "html"), FileTypes.Html);
	}

	public static void verify(File generateFile) {
		verify(new FileApprovalWriter(generateFile), FileTypes.File);
	}

	public static void verify(Image image) {
		// approve(ImageWriter.toBufferedImage(image), createApprovalNamer());
	}

	public static void verify(BufferedImage bufferedImage) {
		verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
	}

	public static void verify(ApprovalWriter writter, ApprovalNamer namer,
			ApprovalFailureReporter reporter) {
		verify(new FileApprover(writter, namer), reporter);
	}

	public static void verify(ApprovalWriter writter, String fileType) {
		verify(writter, createApprovalNamer(), ReporterFactory.get(fileType));
	}

	public static void verifyXml(String xml) throws Exception {
		verify(new ApprovalXmlWriter(xml), FileTypes.Text);
	}

	public static void verify(FileApprover approver,
			ApprovalFailureReporter reporter) {
		try {
			if (!approver.approve()) {
				boolean passed = false;
				if (reporter instanceof ApprovalFailureOverrider) {
					passed = approver
							.askToChangeReceivedToApproved((ApprovalFailureOverrider) reporter);
				}
				if (!passed) {
					approver.reportFailure(reporter);
					approver.fail();
				} else {
					approver.cleanUpAfterSuccess(reporter);
				}
			} else {
				approver.cleanUpAfterSuccess(reporter);
			}
		} catch (Exception e) {
			throw ObjectUtils.throwAsError(e);
		}
	}

	public static void verify(Map map) throws Exception {
		verify(new ApprovalTextWriter(StringUtils.toString(map), "txt"),
				FileTypes.Text);
	}

	public static ApprovalNamer createApprovalNamer() {
		return new JUnitStackTraceNamer();
	}

	private static void approve(BufferedImage bufferedImage, ApprovalNamer namer) {
		verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
	}

//	private static String getFileNameList(List<File> mismatched) {
//		return Query.select(mismatched,
//				new F1<File, String>(mismatched.get(1)) {
//					{
//						ret(a.getName());
//					}
//				}).toString();
//	}
}