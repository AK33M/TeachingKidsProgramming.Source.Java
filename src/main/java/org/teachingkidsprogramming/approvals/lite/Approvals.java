package org.teachingkidsprogramming.approvals.lite;

import org.teachingkidsprogramming.approvals.lite.ReporterFactory.FileTypes;
import org.teachingkidsprogramming.approvals.lite.namer.ApprovalNamer;
import org.teachingkidsprogramming.approvals.lite.namer.JUnitStackTraceNamer;
import org.approvaltests.writers.ApprovalTextWriter;
import org.approvaltests.writers.ImageApprovalWriter;
import org.lambda.functions.Function1;
import org.teachingkidsprogramming.util.ArrayUtils;
import org.teachingkidsprogramming.util.ObjectUtils;
import org.teachingkidsprogramming.util.StringUtils;

import java.awt.image.BufferedImage;
import java.util.Arrays;

public class Approvals {
	public static void verify(String response) throws Exception {
		verify(new ApprovalTextWriter(response, "txt"), FileTypes.Text);
	}

    public static <T> void verifyAll(String header, T[] values,
			Function1<T, String> f1) {
		verifyAll(header, Arrays.asList(values), f1);
	}

	public static <T> void verifyAll(String header, Iterable<T> array,
			Function1<T, String> f1) {
		String text = formatHeader(header) + ArrayUtils.toString(array, f1);
		verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
	}

	private static String formatHeader(String header) {
		return StringUtils.isEmpty(header) ? "" : header + "\r\n\r\n\r\n";
	}

    public static void verifyHtml(String response) throws Exception {
		verify(new ApprovalTextWriter(response, "html"), FileTypes.Html);
	}

    public static void verify(BufferedImage bufferedImage) {
		verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
	}

	public static void verify(ApprovalWriter writter, ApprovalNamer namer,
			ApprovalFailureReporter reporter) {
		verify(new FileApprover(writter, namer), reporter);
	}

	public static void verify(ApprovalWriter writter, String fileType) {
		verify(writter, createApprovalNamer(), ReporterFactory.get(fileType));
	}

    public static void verify(FileApprover approver,
			ApprovalFailureReporter reporter) {
		try {
			if (!approver.approve()) {
				boolean passed = false;
				if (reporter instanceof ApprovalFailureOverrider) {
					passed = approver
							.askToChangeReceivedToApproved((ApprovalFailureOverrider) reporter);
				}
				if (!passed) {
					approver.reportFailure(reporter);
					approver.fail();
				} else {
					approver.cleanUpAfterSuccess(reporter);
				}
			} else {
				approver.cleanUpAfterSuccess(reporter);
			}
		} catch (Exception e) {
			throw ObjectUtils.throwAsError(e);
		}
	}

    public static ApprovalNamer createApprovalNamer() {
		return new JUnitStackTraceNamer();
	}

}