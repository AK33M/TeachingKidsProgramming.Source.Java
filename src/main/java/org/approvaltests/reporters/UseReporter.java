package org.approvaltests.reporters;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import org.teachingkidsprogramming.approvals.lite.ApprovalFailureReporter;

@Retention(RetentionPolicy.RUNTIME)
public @interface UseReporter {
  Class<? extends ApprovalFailureReporter>[] value();
}
