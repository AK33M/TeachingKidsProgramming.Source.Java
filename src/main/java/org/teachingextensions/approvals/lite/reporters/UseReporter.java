package org.teachingextensions.approvals.lite.reporters;

import org.teachingextensions.approvals.lite.ApprovalFailureReporter;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@Retention(RetentionPolicy.RUNTIME)
public @interface UseReporter {
    Class<? extends ApprovalFailureReporter>[] value();
}
