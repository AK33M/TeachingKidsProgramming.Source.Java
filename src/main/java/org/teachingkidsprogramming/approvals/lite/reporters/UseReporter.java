package org.teachingkidsprogramming.approvals.lite.reporters;

import org.teachingkidsprogramming.approvals.lite.ApprovalFailureReporter;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@Retention(RetentionPolicy.RUNTIME)
public @interface UseReporter {
    Class<? extends ApprovalFailureReporter>[] value();
}
