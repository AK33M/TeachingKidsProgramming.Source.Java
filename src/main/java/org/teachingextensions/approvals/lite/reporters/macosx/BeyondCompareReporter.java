package org.teachingextensions.approvals.lite.reporters.macosx;

import org.teachingextensions.approvals.lite.reporters.GenericDiffReporter;

import java.text.MessageFormat;

public class BeyondCompareReporter extends GenericDiffReporter {
  public static final BeyondCompareReporter INSTANCE     = new BeyondCompareReporter();
  static final        String                DIFF_PROGRAM = "/usr/local/bin/bcompare";
  static final        String                MESSAGE      = MessageFormat.format(
      "Unable to find Beyond Compare at {0}", DIFF_PROGRAM);

  public BeyondCompareReporter() {
    super(DIFF_PROGRAM, MESSAGE);
  }
}
