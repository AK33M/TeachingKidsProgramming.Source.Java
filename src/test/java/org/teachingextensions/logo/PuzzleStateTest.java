package org.teachingextensions.logo;

import org.junit.Test;
import org.teachingextensions.approvals.lite.Approvals;
import org.teachingextensions.approvals.lite.reporters.UseReporter;
import org.teachingextensions.approvals.lite.reporters.macosx.BeyondCompareReporter;

import java.util.Stack;

import static org.junit.Assert.*;

@UseReporter(BeyondCompareReporter.class)
public class PuzzleStateTest {

  private static Puzzle getSolvedPuzzle() {
    int[] solved = {0, 1, 2, 3, 4, 5, 6, 7, 8};
    return new Puzzle(solved);
  }

  private static Puzzle getPuzzle(int blank) {
    return getSolvedPuzzle().swapBlank(blank);
  }

  /**
   * The state can tell you what moves are allowed from cell to cell
   */
  @Test
  public void state_has_move_directions() {
    Approvals.verifyAll("Directions", PuzzleState.Direction.values());
  }

  /**
   * The state can tell you if it is a solution to the puzzle
   */
  @Test
  public void state_can_be_solution() {
    Puzzle p = getSolvedPuzzle();
    PuzzleState s = new PuzzleState(p);
    assertTrue(s.isSolution());
  }

  /**
   * The state can tell you all possible next moves for its puzzle.
   */
  @Test
  public void tells_you_next_moves() {
    PuzzleState s = new PuzzleState(getSolvedPuzzle());
    Approvals.verifyAll("Branches for " + s, s.getBranches());
  }

  /**
   * Or we can see the next moves from the top right corner
   */
  @Test
  public void moves_from_top_right() {
    Puzzle p = getPuzzle(2);
    PuzzleState s = new PuzzleState(p);
    Approvals.verifyAll("Branches for " + s, s.getBranches());
  }

  /**
   * Or we can see the next moves from the center
   */
  @Test
  public void moves_from_center() {
    Puzzle p = getPuzzle(4);
    PuzzleState s = new PuzzleState(p);
    Approvals.verifyAll("Branches for " + s, s.getBranches());
  }

  /**
   * Finally, lets see what moves we can do from top center.
   */
  @Test
  public void moves_from_top_center() {
    PuzzleState s = new PuzzleState(getPuzzle(1));
    Approvals.verifyAll("Branches for " + s, s.getBranches());
  }

  /**
   * The actual cost is zero when there is no history
   */
  @Test
  public void actual_cost() {
    PuzzleState s = new PuzzleState(getSolvedPuzzle());
    assertEquals(0, s.getActualCost());
  }

  /**
   * The actual cost is the number of steps in the the history
   */
  @Test
  public void actual_cost_is_actual_steps() {
    Stack<PuzzleState.Direction> history = new Stack<>();
    history.add(PuzzleState.Direction.Right);
    history.add(PuzzleState.Direction.Right);
    PuzzleState s = new PuzzleState(getSolvedPuzzle(), history);
    assertEquals(2, s.getActualCost());
  }

  /**
   * Is not equal to another state when the puzzles are different
   */
  @Test
  public void not_equal_to_state_with_different_puzzle() {
    PuzzleState a = new PuzzleState(getSolvedPuzzle());
    PuzzleState b = new PuzzleState(getPuzzle(2));
    assertNotEquals(a, b);
  }

  /**
   * Is equal to another state when the puzzles are the same
   */
  @Test
  public void equal_to_state_with_same_puzzle() {
    PuzzleState a = new PuzzleState(getPuzzle(1));
    PuzzleState b = new PuzzleState(getPuzzle(1));
    assertTrue(a.equals(b));
  }

  /**
   * Is equal to state with same puzzle even when history differs
   */
  @Test
  public void equal_even_with_different_history() {
    Stack<PuzzleState.Direction> history = new Stack<>();
    history.add(PuzzleState.Direction.Right);
    PuzzleState a = new PuzzleState(getPuzzle(3), history);
    PuzzleState b = new PuzzleState(getPuzzle(3));
    assertEquals(a, b);
  }
}
