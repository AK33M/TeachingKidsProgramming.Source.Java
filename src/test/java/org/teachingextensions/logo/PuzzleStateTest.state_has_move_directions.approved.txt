Directions


{Left = -1}
{Right = 1}
{Up = -3}
{Down = 3}
